File: DynamicKey/AgoraDynamicKey/python3/src/RtcTokenBuilder.py
# -*- coding: utf-8 -*-
__copyright__ = "Copyright (c) 2014-2017 Agora.io, Inc."


from .AccessToken import *


warnings.warn('The RtcTokenBuilder module is deprecated', DeprecationWarning)


Role_Attendee = 0  # depreated, same as publisher
Role_Publisher = 1  # for live broadcaster
Role_Subscriber = 2  # default, for live audience
Role_Admin = 101  # deprecated, same as publisher


class RtcTokenBuilder:
    # appID: The App ID issued to you by Agora. Apply for a new App ID from 
    #        Agora Dashboard if it is missing from your kit. See Get an App ID.
    # appCertificate:	Certificate of the application that you registered in 
    #                  the Agora Dashboard. See Get an App Certificate.
    # channelName:Unique channel name for the AgoraRTC session in the string format
    # uid: User ID. A 32-bit unsigned integer with a value ranging from 
    #      1 to (232-1). optionalUid must be unique.
    # role: Role_Publisher = 1: A broadcaster (host) in a live-broadcast profile.
    #       Role_Subscriber = 2: (Default) A audience in a live-broadcast profile.
    # privilegeExpireTs: represented by the number of seconds elapsed since 
    #                    1/1/1970. If, for example, you want to access the
    #                    Agora Service within 10 minutes after the token is 
    #                    generated, set expireTimestamp as the current 
    #                    timestamp + 600 (seconds)./
    @staticmethod
    def buildTokenWithUid(appId, appCertificate, channelName, uid, role, privilegeExpiredTs):
        return RtcTokenBuilder.buildTokenWithAccount(appId, appCertificate, channelName, uid, role, privilegeExpiredTs)

    # appID: The App ID issued to you by Agora. Apply for a new App ID from 
    #        Agora Dashboard if it is missing from your kit. See Get an App ID.
    # appCertificate:	Certificate of the application that you registered in 
    #                  the Agora Dashboard. See Get an App Certificate.
    # channelName:Unique channel name for the AgoraRTC session in the string format
    # userAccount: The user account. 
    # role: Role_Publisher = 1: A broadcaster (host) in a live-broadcast profile.
    #       Role_Subscriber = 2: (Default) A audience in a live-broadcast profile.
    # privilegeExpireTs: represented by the number of seconds elapsed since 
    #                    1/1/1970. If, for example, you want to access the
    #                    Agora Service within 10 minutes after the token is 
    #                    generated, set expireTimestamp as the current 
    @staticmethod
    def buildTokenWithAccount(appId, appCertificate, channelName, account, role, privilegeExpiredTs):
        token = AccessToken(appId, appCertificate, channelName, account)
        token.addPrivilege(kJoinChannel, privilegeExpiredTs)
        if (role == Role_Attendee) | (role == Role_Admin) | (role == Role_Publisher):
            token.addPrivilege(kPublishVideoStream, privilegeExpiredTs)
            token.addPrivilege(kPublishAudioStream, privilegeExpiredTs)
            token.addPrivilege(kPublishDataStream, privilegeExpiredTs)
        return token.build()

    '''
    * Generates an RTC token with the specified privilege.
    *
    * This method supports generating a token with the following privileges:
    * - Joining an RTC channel.
    * - Publishing audio in an RTC channel.
    * - Publishing video in an RTC channel.
    * - Publishing data streams in an RTC channel.
    *
    * The privileges for publishing audio, video, and data streams in an RTC channel apply only if you have
    * enabled co-host authentication.
    *
    * A user can have multiple privileges. Each privilege is valid for a maximum of 24 hours.
    * The SDK triggers the onTokenPrivilegeWillExpire and onRequestToken callbacks when the token is about to expire
    * or has expired. The callbacks do not report the specific privilege affected, and you need to maintain
    * the respective timestamp for each privilege in your app logic. After receiving the callback, you need
    * to generate a new token, and then call renewToken to pass the new token to the SDK, or call joinChannel to re-join
    * the channel.
    *
    * @note
    * Agora recommends setting a reasonable timestamp for each privilege according to your scenario.
    * Suppose the expiration timestamp for joining the channel is set earlier than that for publishing audio.
    * When the token for joining the channel expires, the user is immediately kicked off the RTC channel
    * and cannot publish any audio stream, even though the timestamp for publishing audio has not expired.
    *
    * @param appId The App ID of your Agora project.
    * @param appCertificate The App Certificate of your Agora project.
    * @param channelName The unique channel name for the Agora RTC session in string format. The string length must be less than 64 bytes. The channel name may contain the following characters:
    * - All lowercase English letters: a to z.
    * - All uppercase English letters: A to Z.
    * - All numeric characters: 0 to 9.
    * - The space character.
    * - "!", "#", "$", "%", "&", "(", ")", "+", "-", ":", ";", "<", "=", ".", ">", "?", "@", "[", "]", "^", "_", " {", "}", "|", "~", ",".
    * @param uid The user ID. A 32-bit unsigned integer with a value range from 1 to (232 - 1). It must be unique. Set uid as 0, if you do not want to authenticate the user ID, that is, any uid from the app client can join the channel.
    * @param joinChannelPrivilegeExpiredTs The Unix timestamp when the privilege for joining the channel expires, represented
    * by the sum of the current timestamp plus the valid time period of the token. For example, if you set joinChannelPrivilegeExpiredTs as the
    * current timestamp plus 600 seconds, the token expires in 10 minutes.
    * @param pubAudioPrivilegeExpiredTs The Unix timestamp when the privilege for publishing audio expires, represented
    * by the sum of the current timestamp plus the valid time period of the token. For example, if you set pubAudioPrivilegeExpiredTs as the
    * current timestamp plus 600 seconds, the token expires in 10 minutes. If you do not want to enable this privilege,
    * set pubAudioPrivilegeExpiredTs as the current Unix timestamp.
    * @param pubVideoPrivilegeExpiredTs The Unix timestamp when the privilege for publishing video expires, represented
    * by the sum of the current timestamp plus the valid time period of the token. For example, if you set pubVideoPrivilegeExpiredTs as the
    * current timestamp plus 600 seconds, the token expires in 10 minutes. If you do not want to enable this privilege,
    * set pubVideoPrivilegeExpiredTs as the current Unix timestamp.
    * @param pubDataStreamPrivilegeExpiredTs The Unix timestamp when the privilege for publishing data streams expires, represented
    * by the sum of the current timestamp plus the valid time period of the token. For example, if you set pubDataStreamPrivilegeExpiredTs as the
    * current timestamp plus 600 seconds, the token expires in 10 minutes. If you do not want to enable this privilege,
    * set pubDataStreamPrivilegeExpiredTs as the current Unix timestamp.
    * @return The new Token
    '''

    @staticmethod
    def buildTokenWithUidAndPrivilege(appId, appCertificate, channelName, uid, joinChannelPrivilegeExpiredTs,
                                              pubAudioPrivilegeExpiredTs, pubVideoPrivilegeExpiredTs,
                                              pubDataStreamPrivilegeExpiredTs):
        return RtcTokenBuilder.buildTokenWithUserAccountAndPrivilege(appId, appCertificate, channelName, uid,
                                                     joinChannelPrivilegeExpiredTs, pubAudioPrivilegeExpiredTs,
                                                     pubVideoPrivilegeExpiredTs, pubDataStreamPrivilegeExpiredTs)

    '''
    * Generates an RTC token with the specified privilege.
    *
    * This method supports generating a token with the following privileges:
    * - Joining an RTC channel.
    * - Publishing audio in an RTC channel.
    * - Publishing video in an RTC channel.
    * - Publishing data streams in an RTC channel.
    *
    * The privileges for publishing audio, video, and data streams in an RTC channel apply only if you have
    * enabled co-host authentication.
    *
    * A user can have multiple privileges. Each privilege is valid for a maximum of 24 hours.
    * The SDK triggers the onTokenPrivilegeWillExpire and onRequestToken callbacks when the token is about to expire
    * or has expired. The callbacks do not report the specific privilege affected, and you need to maintain
    * the respective timestamp for each privilege in your app logic. After receiving the callback, you need
    * to generate a new token, and then call renewToken to pass the new token to the SDK, or call joinChannel to re-join
    * the channel.
    *
    * @note
    * Agora recommends setting a reasonable timestamp for each privilege according to your scenario.
    * Suppose the expiration timestamp for joining the channel is set earlier than that for publishing audio.
    * When the token for joining the channel expires, the user is immediately kicked off the RTC channel
    * and cannot publish any audio stream, even though the timestamp for publishing audio has not expired.
    *
    * @param appId The App ID of your Agora project.
    * @param appCertificate The App Certificate of your Agora project.
    * @param channelName The unique channel name for the Agora RTC session in string format. The string length must be less than 64 bytes. The channel name may contain the following characters:
    * - All lowercase English letters: a to z.
    * - All uppercase English letters: A to Z.
    * - All numeric characters: 0 to 9.
    * - The space character.
    * - "!", "#", "$", "%", "&", "(", ")", "+", "-", ":", ";", "<", "=", ".", ">", "?", "@", "[", "]", "^", "_", " {", "}", "|", "~", ",".
    * @param userAccount The user account.
    * @param joinChannelPrivilegeExpiredTs The Unix timestamp when the privilege for joining the channel expires, represented
    * by the sum of the current timestamp plus the valid time period of the token. For example, if you set joinChannelPrivilegeExpiredTs as the
    * current timestamp plus 600 seconds, the token expires in 10 minutes.
    * @param pubAudioPrivilegeExpiredTs The Unix timestamp when the privilege for publishing audio expires, represented
    * by the sum of the current timestamp plus the valid time period of the token. For example, if you set pubAudioPrivilegeExpiredTs as the
    * current timestamp plus 600 seconds, the token expires in 10 minutes. If you do not want to enable this privilege,
    * set pubAudioPrivilegeExpiredTs as the current Unix timestamp.
    * @param pubVideoPrivilegeExpiredTs The Unix timestamp when the privilege for publishing video expires, represented
    * by the sum of the current timestamp plus the valid time period of the token. For example, if you set pubVideoPrivilegeExpiredTs as the
    * current timestamp plus 600 seconds, the token expires in 10 minutes. If you do not want to enable this privilege,
    * set pubVideoPrivilegeExpiredTs as the current Unix timestamp.
    * @param pubDataStreamPrivilegeExpiredTs The Unix timestamp when the privilege for publishing data streams expires, represented
    * by the sum of the current timestamp plus the valid time period of the token. For example, if you set pubDataStreamPrivilegeExpiredTs as the
    * current timestamp plus 600 seconds, the token expires in 10 minutes. If you do not want to enable this privilege,
    * set pubDataStreamPrivilegeExpiredTs as the current Unix timestamp.
    * @return The new Token
    '''

    @staticmethod
    def buildTokenWithUserAccountAndPrivilege(appId, appCertificate, channelName, account,
                                                  joinChannelPrivilegeExpiredTs,
                                                  pubAudioPrivilegeExpiredTs, pubVideoPrivilegeExpiredTs,
                                                  pubDataStreamPrivilegeExpiredTs):
        token = AccessToken(appId, appCertificate, channelName, account)
        token.addPrivilege(kJoinChannel, joinChannelPrivilegeExpiredTs)
        token.addPrivilege(kPublishAudioStream, pubAudioPrivilegeExpiredTs)
        token.addPrivilege(kPublishVideoStream, pubVideoPrivilegeExpiredTs)
        token.addPrivilege(kPublishDataStream, pubDataStreamPrivilegeExpiredTs)
        return token.build()

File: DynamicKey/AgoraDynamicKey/dart/lib/src/rtc_token_builder.dart
import 'access_token.dart';

/// Defines the RTC role used to select token privileges.
enum RtcRole {
  /// Recommended for voice/video calls and live broadcasts when
  /// [co-host authentication](https://docs.agora.io/en/video-calling/get-started/authentication-workflow?#co-host-token-authentication)
  /// is not required.
  publisher,

  /// Use only when
  /// [co-host authentication](https://docs.agora.io/en/video-calling/get-started/authentication-workflow?#co-host-token-authentication)
  /// is required. Contact Agora support to enable it; otherwise this role has
  /// the same privileges as the publisher role.
  subscriber,
}

/// Builds Token007 tokens for RTC and combined RTC/RTM services.
class RtcTokenBuilder {
  /// Builds an RTC token for a numeric user ID.
  ///
  /// [appId] is the App ID issued by Agora; create one in Agora Dashboard if needed.
  /// [appCertificate] is the certificate of the application registered in Agora Dashboard.
  /// [channelName] is the unique channel name for the RTC session.
  /// [uid] is a unique unsigned 32-bit user ID; use `0` to allow any user ID.
  /// [role] is [RtcRole.publisher] for a broadcaster or [RtcRole.subscriber] for an audience member.
  /// [tokenExpireSeconds] is the number of seconds from now until the token expires; use `600` for 10 minutes.
  /// [privilegeExpireSeconds] is the number of seconds from now until the RTC privileges expire; it defaults to the token lifetime.
  ///
  /// Returns the generated RTC token.
  static String buildTokenWithUid({
    required String appId,
    required String appCertificate,
    required String channelName,
    required int uid,
    required int tokenExpireSeconds,
    RtcRole role = RtcRole.publisher,
    int? privilegeExpireSeconds,
  }) {
    return buildTokenWithUserAccount(
      appId: appId,
      appCertificate: appCertificate,
      channelName: channelName,
      account: uid == 0 ? '' : uid.toString(),
      role: role,
      tokenExpireSeconds: tokenExpireSeconds,
      privilegeExpireSeconds: privilegeExpireSeconds ?? tokenExpireSeconds,
    );
  }

  /// Builds an RTC token for a user account.
  ///
  /// [appId] is the App ID issued by Agora; create one in Agora Dashboard if needed.
  /// [appCertificate] is the certificate of the application registered in Agora Dashboard.
  /// [channelName] is the unique channel name for the RTC session.
  /// [account] is the user's account and must not exceed 255 bytes.
  /// [role] is [RtcRole.publisher] for a broadcaster or [RtcRole.subscriber] for an audience member.
  /// [tokenExpireSeconds] is the number of seconds from now until the token expires; use `600` for 10 minutes.
  /// [privilegeExpireSeconds] is the number of seconds from now until the RTC privileges expire; it defaults to the token lifetime.
  ///
  /// Returns the generated RTC token.
  static String buildTokenWithUserAccount({
    required String appId,
    required String appCertificate,
    required String channelName,
    required String account,
    required int tokenExpireSeconds,
    RtcRole role = RtcRole.publisher,
    int? privilegeExpireSeconds,
  }) {
    final privilegeExpire = privilegeExpireSeconds ?? tokenExpireSeconds;
    final token = AccessToken.create(
      appId,
      appCertificate,
      expire: tokenExpireSeconds,
    );

    final serviceRtc = ServiceRtc(channelName, account);
    serviceRtc.addPrivilege(
      ServiceRtc.privilegeJoinChannel,
      privilegeExpire,
    );
    if (role == RtcRole.publisher) {
      serviceRtc.addPrivilege(
        ServiceRtc.privilegePublishAudioStream,
        privilegeExpire,
      );
      serviceRtc.addPrivilege(
        ServiceRtc.privilegePublishVideoStream,
        privilegeExpire,
      );
      serviceRtc.addPrivilege(
        ServiceRtc.privilegePublishDataStream,
        privilegeExpire,
      );
    }
    token.addService(serviceRtc);

    return token.build();
  }

  /// Builds an RTC token for a user account using the original Dart API name.
  ///
  /// [appId] is the App ID issued by Agora; create one in Agora Dashboard if needed.
  /// [appCertificate] is the certificate of the application registered in Agora Dashboard.
  /// [channelName] is the unique channel name for the RTC session.
  /// [account] is the user's account and must not exceed 255 bytes.
  /// [role] is [RtcRole.publisher] for a broadcaster or [RtcRole.subscriber] for an audience member.
  /// [tokenExpireSeconds] is the number of seconds from now until the token expires; use `600` for 10 minutes.
  /// [privilegeExpireSeconds] is the number of seconds from now until the RTC privileges expire; it defaults to the token lifetime.
  ///
  /// This compatibility alias behaves like [buildTokenWithUserAccount].
  static String buildTokenWithAccount({
    required String appId,
    required String appCertificate,
    required String channelName,
    required String account,
    required int tokenExpireSeconds,
    RtcRole role = RtcRole.publisher,
    int? privilegeExpireSeconds,
  }) {
    return buildTokenWithUserAccount(
      appId: appId,
      appCertificate: appCertificate,
      channelName: channelName,
      account: account,
      role: role,
      tokenExpireSeconds: tokenExpireSeconds,
      privilegeExpireSeconds: privilegeExpireSeconds,
    );
  }

  /// Builds an RTC token with independent privilege expiration values for a
  /// numeric user ID.
  ///
  /// The token supports channel join and audio, video, and data publishing
  /// privileges. Publishing privileges apply only when co-host authentication
  /// is enabled. Each privilege can be valid for up to 24 hours.
  ///
  /// [appId] is the App ID issued by Agora.
  /// [appCertificate] is the App Certificate of the Agora project.
  /// [channelName] is the unique RTC channel name and must be less than 64 bytes.
  /// [uid] is a unique unsigned 32-bit user ID; use `0` to allow any user ID.
  /// [tokenExpireSeconds] is the number of seconds from now until the token expires; use `600` for 10 minutes.
  /// [joinChannelPrivilegeExpire] is the number of seconds from now until the channel join privilege expires.
  /// [publishAudioPrivilegeExpire] is the number of seconds from now until the audio publishing privilege expires.
  /// [publishVideoPrivilegeExpire] is the number of seconds from now until the video publishing privilege expires.
  /// [publishDataPrivilegeExpire] is the number of seconds from now until the data publishing privilege expires.
  ///
  /// Returns the generated RTC token.
  static String buildTokenWithUidAndPrivilege({
    required String appId,
    required String appCertificate,
    required String channelName,
    required int uid,
    required int tokenExpireSeconds,
    required int joinChannelPrivilegeExpire,
    required int publishAudioPrivilegeExpire,
    required int publishVideoPrivilegeExpire,
    required int publishDataPrivilegeExpire,
  }) {
    return buildTokenWithUserAccountAndPrivilege(
      appId: appId,
      appCertificate: appCertificate,
      channelName: channelName,
      account: uid == 0 ? '' : uid.toString(),
      tokenExpireSeconds: tokenExpireSeconds,
      joinChannelPrivilegeExpire: joinChannelPrivilegeExpire,
      publishAudioPrivilegeExpire: publishAudioPrivilegeExpire,
      publishVideoPrivilegeExpire: publishVideoPrivilegeExpire,
      publishDataPrivilegeExpire: publishDataPrivilegeExpire,
    );
  }

  /// Builds an RTC token with independent privilege expiration values for a
  /// user account.
  ///
  /// The token supports channel join and audio, video, and data publishing
  /// privileges. Publishing privileges apply only when co-host authentication
  /// is enabled. Each privilege can be valid for up to 24 hours.
  ///
  /// [appId] is the App ID issued by Agora.
  /// [appCertificate] is the App Certificate of the Agora project.
  /// [channelName] is the unique RTC channel name and must be less than 64 bytes.
  /// [account] is the user account.
  /// [tokenExpireSeconds] is the number of seconds from now until the token expires; use `600` for 10 minutes.
  /// [joinChannelPrivilegeExpire] is the number of seconds from now until the channel join privilege expires.
  /// [publishAudioPrivilegeExpire] is the number of seconds from now until the audio publishing privilege expires.
  /// [publishVideoPrivilegeExpire] is the number of seconds from now until the video publishing privilege expires.
  /// [publishDataPrivilegeExpire] is the number of seconds from now until the data publishing privilege expires.
  ///
  /// Returns the generated RTC token.
  static String buildTokenWithUserAccountAndPrivilege({
    required String appId,
    required String appCertificate,
    required String channelName,
    required String account,
    required int tokenExpireSeconds,
    required int joinChannelPrivilegeExpire,
    required int publishAudioPrivilegeExpire,
    required int publishVideoPrivilegeExpire,
    required int publishDataPrivilegeExpire,
  }) {
    final token = AccessToken.create(
      appId,
      appCertificate,
      expire: tokenExpireSeconds,
    );

    final serviceRtc = ServiceRtc(channelName, account);
    serviceRtc.addPrivilege(
      ServiceRtc.privilegeJoinChannel,
      joinChannelPrivilegeExpire,
    );
    serviceRtc.addPrivilege(
      ServiceRtc.privilegePublishAudioStream,
      publishAudioPrivilegeExpire,
    );
    serviceRtc.addPrivilege(
      ServiceRtc.privilegePublishVideoStream,
      publishVideoPrivilegeExpire,
    );
    serviceRtc.addPrivilege(
      ServiceRtc.privilegePublishDataStream,
      publishDataPrivilegeExpire,
    );
    token.addService(serviceRtc);

    return token.build();
  }

  /// Builds one Token007 token containing RTC and RTM services for the same
  /// account.
  ///
  /// [appId] is the App ID issued by Agora; create one in Agora Dashboard if needed.
  /// [appCertificate] is the certificate of the application registered in Agora Dashboard.
  /// [channelName] is the unique channel name for the RTC session.
  /// [account] is the user's account and must not exceed 255 bytes.
  /// [role] is [RtcRole.publisher] for a broadcaster or [RtcRole.subscriber] for an audience member.
  /// [tokenExpireSeconds] is the number of seconds from now until the token expires; use `600` for 10 minutes.
  /// [privilegeExpireSeconds] is the number of seconds from now until the RTC privileges expire; use `600` for 10 minutes.
  ///
  /// Returns the generated RTC and RTM token.
  static String buildTokenWithRtm({
    required String appId,
    required String appCertificate,
    required String channelName,
    required String account,
    required RtcRole role,
    required int tokenExpireSeconds,
    required int privilegeExpireSeconds,
  }) {
    return buildTokenWithRtm2(
      appId: appId,
      appCertificate: appCertificate,
      channelName: channelName,
      rtcAccount: account,
      rtcRole: role,
      rtcTokenExpireSeconds: tokenExpireSeconds,
      joinChannelPrivilegeExpire: privilegeExpireSeconds,
      publishAudioPrivilegeExpire: privilegeExpireSeconds,
      publishVideoPrivilegeExpire: privilegeExpireSeconds,
      publishDataPrivilegeExpire: privilegeExpireSeconds,
      rtmUserId: account,
      rtmTokenExpireSeconds: tokenExpireSeconds,
    );
  }

  /// Builds one Token007 token containing RTC and RTM services with separate
  /// accounts and expiration values.
  ///
  /// [appId] is the App ID issued by Agora; create one in Agora Dashboard if needed.
  /// [appCertificate] is the certificate of the application registered in Agora Dashboard.
  /// [channelName] is the unique channel name for the RTC session.
  /// [rtcAccount] is the RTC user's account and must not exceed 255 bytes.
  /// [rtcRole] is [RtcRole.publisher] for a broadcaster or [RtcRole.subscriber] for an audience member.
  /// [rtcTokenExpireSeconds] is the number of seconds from now until the RTC token expires; use `600` for 10 minutes.
  /// [joinChannelPrivilegeExpire] is the number of seconds from now until the channel join privilege expires.
  /// [publishAudioPrivilegeExpire] is the number of seconds from now until the audio publishing privilege expires.
  /// [publishVideoPrivilegeExpire] is the number of seconds from now until the video publishing privilege expires.
  /// [publishDataPrivilegeExpire] is the number of seconds from now until the data publishing privilege expires.
  /// [rtmUserId] is the RTM user's account and must not exceed 255 bytes.
  /// [rtmTokenExpireSeconds] is the number of seconds from now until the RTM token expires; use `600` for 10 minutes.
  ///
  /// Returns the generated RTC and RTM token.
  static String buildTokenWithRtm2({
    required String appId,
    required String appCertificate,
    required String channelName,
    required String rtcAccount,
    required RtcRole rtcRole,
    required int rtcTokenExpireSeconds,
    required int joinChannelPrivilegeExpire,
    required int publishAudioPrivilegeExpire,
    required int publishVideoPrivilegeExpire,
    required int publishDataPrivilegeExpire,
    required String rtmUserId,
    required int rtmTokenExpireSeconds,
  }) {
    final token = AccessToken.create(
      appId,
      appCertificate,
      expire: rtcTokenExpireSeconds,
    );

    final serviceRtc = ServiceRtc(channelName, rtcAccount);
    serviceRtc.addPrivilege(
      ServiceRtc.privilegeJoinChannel,
      joinChannelPrivilegeExpire,
    );
    if (rtcRole == RtcRole.publisher) {
      serviceRtc.addPrivilege(
        ServiceRtc.privilegePublishAudioStream,
        publishAudioPrivilegeExpire,
      );
      serviceRtc.addPrivilege(
        ServiceRtc.privilegePublishVideoStream,
        publishVideoPrivilegeExpire,
      );
      serviceRtc.addPrivilege(
        ServiceRtc.privilegePublishDataStream,
        publishDataPrivilegeExpire,
      );
    }
    token.addService(serviceRtc);

    token.addService(
      ServiceRtm(rtmUserId)
        ..addPrivilege(
          ServiceRtm.privilegeLogin,
          rtmTokenExpireSeconds,
        ),
    );
    return token.build();
  }
}
